apps/src/templates/teacherDashboard/teacherSectionsRedux.js

```diff
@@ -92,16 +159,35 @@ export const assignments = state =>
  */
 export const currentAssignmentIndex = (state, sectionId) => {
   const section = state.sections[sectionId];
-  const assignmentList = assignments(state);
-  const target = {
-    courseId: section.courseId,
-    scriptId: section.courseId ? null : section.scriptId
-  };
 
-  // Find an assignment with the appropriate id
-  const assignmentIndex = assignmentList.findIndex(assignment => (
-    assignment.courseId === target.courseId && assignment.scriptId === target.scriptId
-  ));
+  const assignmentIndex = getAssignmentIndex(assignments(state))(
+    section.courseId, section.scriptId);
 
   return assignmentIndex === -1 ? null : assignmentIndex;
 };
+
+/**
+ * Find an assignment with the appropriate id. If both courseId and scriptId are
+ * non-null, look only at the courseId.
+ * This is memoized at two separate levels because JS doesnt have any good way
+ * to have a cache key consisting of an object (assignmentList) and two strings.
+ * @param {object[]} assignmentList - array of valid courses/scripts
+ * @param {string} courseId - course id of the assignment we're looking for
+ * @param {string} scriptId - script id of the assignment we're looking for
+ */
+const getAssignmentIndex = _.memoize(assignmentList => (
+  _.memoize((courseId, scriptId) => (
+    assignmentList.findIndex(assignment =>
+      assignment.courseId === courseId &&
+      (courseId || assignment.scriptId === scriptId)
+    )
+  ), (courseId, scriptId) => `${courseId}_${scriptId}`)
+));
```

Comment:
This works, but I don't think that it's easy to understand. I think that what I should do instead is as follows:
- Combine `setValidCourse` and `setValidScripts` to be `setValidAssignments(validCourses, validScripts)`.
- In the reducer, instead of storing these as a list, store them as a mapping from key to assignment (where key is something like `${courseId}_${scriptId}`
- Also generate our path at this time
- Now it's trivial to get from section to assignment (via courseId/scriptId).
- `AssignmentSelector` will still need a list of assignments, where order matters. We could accomplish that be either (a) storing a list of assignments next to our mapping or (b) storing a list of keys, and then generating the list of assignments as need (this is probably better).